### DISCUSSION TITLE
Loading metadata from cache only (on production)

### DISCUSSION THREAD

**Discussion by JanJakes** [mikro-orm/mikro-orm#2103] [Q&A]
Is it possible to load entity metadata only from pregenerated cache without the need for `.d.ts`/`.ts` files?

This would also allow having `ts-morph` (`@mikro-orm/reflection`) as a dev dependency only, given that metadata are dumped and packed with the production build. Given the size of ts-morph this can be a nice benefit as well.

The main benefit, though, may be the possibility to transpile code using tools such as SWC or esbuild that don't produce `d.ts` files.

I'm not fully sure whether this is somehow achievable today or if it should be a feature suggestion.

**Comment by B4nan** [comment]
Yes, you can use the CLI to warm up production cache:

`npx mikro-orm cache:generate`

But you still probably need the `TsMorphMetadataProvider` class, or better say its `loadFromCache` method implementation, as that is different from `ReflectMetadataProvider`. Not 100% sure tho, and if so, you could just implement custom metadata provider that will have only this particular method copied from the `TsMorphMetadataProvider` one.

Another way would be to nuke the dependency via webpack/null-loader I guess. 

edit: the idea with custom provider would be probably a bit more complicated - you would have to use the `TsMorphMetadataProvider` with the `cache:generate` command, and then in production when running the app, you'd use the custom one just with the `loadFromCache` method

**Reply by JanJakes** [reply]
Thanks! I'll try it out when I have a bit of time and will let you know.

**Reply by JanJakes** [reply]
I tried it. And got quite confused, initially 🤔

I added a custom metadata provider implementation that just copies the `loadFromCache` method from `TsMorphMetadataProvider`, as you said. I initialize it conditionally:
```ts
const metadataProvider = useOnlyCachedMetadata
  ? (await import('./CachedMetadataProvider')).CachedMetadataProvider
  : (await import('@mikro-orm/reflection')).TsMorphMetadataProvider;
```

Then I run `nest:build`, made sure to exclude `.d.ts` for entities (`rm -rf dist/entities/*.d.ts`), made sure I have no cache generated (deleted `./temp`) and didn't dump any (yet), run `yarn workspaces focus @myproject/api --production` to have production deps only, and tried to run `node dist/main.js`. That's where I expected it to fail on the missing metadata. But it didn't 🙀

What's also weird is I get `[discovery] ORM entity discovery started, using ReflectMetadataProvider` so it doesn't seem to use my `CachedMetadataProvider` (although I verified it gets imported at least).

I pass entities as an array of classes (without discovery), the important parts of the config are as follows (the `CachedMetadataProvider` is now hardcoded):
```ts
metadataProvider: (await import('./CachedMetadataProvider')).CachedMetadataProvider,
baseDir: __dirname,
cache: { enabled: true },
entities, // array of imported entity classes
discovery: { disableDynamicFileAccess: true },
```

We omit many types in the code indeed, such as:
```ts
@Property({ columnType: 'timestamptz' })
createdAt = new Date();
```
Although it looks like we always specify explicit type in `OneToMany` entities and when double-checking why, it seems to me that decorator simply requires that (1-3 params). But we also have enums such as:
```
@Enum(() => MyEnum)
provider: MyEnum;
```

---

So I started to think our code is kind of `ReflectMetadataProvider`-compatible as well. Dumping the metadata in the production build outputs `type: 'object'` for `createdAt` and nullability seems to respect only what's in the decorators. But what about enums, for instance? Interestingly, the code seems to work. But can this have some runtime side-effects? (That metadata is resolved differently than in dev?)

And there seems to be a "fallback" to `ReflectMetadataProvider`... So I digged deeper and found out it's the `discovery: { disableDynamicFileAccess: true }` because of [this](https://github.com/mikro-orm/mikro-orm/blob/9411f4a33c604c044ff2de48fa73a17445bf81f2/packages/core/src/MikroORM.ts#L56-L57). I'm not sure what would be the best solution or a fix here, maybe set those only if nothing else is set in the config?

Fixing that, I finally got somewhere with the following code:
```ts
import { EntityMetadata, MetadataProvider, Utils } from '@mikro-orm/core';

export class CachedMetadataProvider extends MetadataProvider {
  useCache(): boolean {
    return true;
  }

  // Taken from MikroORM's TsMorphMetadataProvider
  loadFromCache(meta: EntityMetadata, cache: EntityMetadata): void {
    Object.values(cache.properties).forEach((prop) => {
      if (prop.customType) {
        prop.customType = meta.properties[prop.name].customType;
      }
    });

    Utils.merge(meta, cache);
  }

  async loadEntityMetadata(meta: EntityMetadata, name: string): Promise<void> {
    return;
  }
}
```

But:
1. The dumped cache files have `.ts`, while production build looks for `.js`.
2. The dumped files have `"origin":"./src/entities/MyEntity.ts"` while production wants `../entities/MyEntity.js`.
3. The dumped files may have a different `hash` maybe? (If its computation includes the path.)

(Is that because `yarn mikro-orm cache:generate` may use `ts-node`?)

As a result, the production app builds the cache again but that's done from the decorators only with data missing. And now this fails on enums with `Cannot read property 'toLowerCase' of undefined`. Quite interestingly, because with the previous  "fallback" to `ReflectMetadataProvider` this wasn't an issue (enums somehow worked).

So what could be done at the moment, I guess, is to use the production build to dump metadata when dev deps are still installed somehow? It's getting a bit complicated... 🤔

**Reply by derolf** [reply]
Did you solve this issue?

**Reply by derolf** [reply]
It would be nice to have a “force”-cache option.

**Reply by B4nan** [reply]
> It would be nice to have a “force”-cache option.

That's basically just a validation for cache-miss, you can already implement that yourself by extending `FileCacheAdapter` and overriding the `get` method.

Or maybe better alternative - you can throw from `loadEntityMetadata` method, will work the same. Both cache adapter and metadata provider classes are configurable.

**Reply by derolf** [reply]
Yeah, I looked into the code and have an idea what I need to do. However, would be nicer if mikroom would support it ootb instead of building a “workaround”…

**Comment by derolf** [comment]
```TypeScript
/**
 * The hash is always 0.
 */
export default class ForceFileCacheAdapter extends FileCacheAdapter {
  constructor(
    options: {
      cacheDir: string;
    },
    baseDir: string,
    pretty?: boolean,
  ) {
    super(options, baseDir, pretty);
    (this as AnyObject).getHash = () => 0;
  }
}
```

**Reply by B4nan** [reply]
This seems very wrong to me. Using this locally will break the cache invalidation (and I'd be very scared to use schema diffing/migrations then), and in production you loose the validation, so you can end up with invalid cache being used, so wrong metadata, hence wrongly functioning ORM.

**Reply by derolf** [reply]
okay, so better to have a specific FileCacheAdapter that just overrides get?

**Reply by B4nan** [reply]
I am not sure what you are trying to solve by that. If you are up to the cache-miss validation, I already told you what to override how. I would go with the `loadEntityMetadata` method that throws. But keep in mind you can use such adapter only once the cache is generated (so to generate it you'd still use the ts-morph one).

**Reply by derolf** [reply]
It feels very shaky. Would it work with the vanilla reflection provider instead of tsmorph?

**Reply by B4nan** [reply]
You don't need caching if you don't use ts-morph.

**Reply by derolf** [reply]
So if we migrate to ReflectMetadataProvider, everything is good?

**Reply by B4nan** [reply]
I have no idea what problems are you trying to solve. If you are up to speeding up the init phase, than simple CLI call to warm up the cache during build time is enough, as I already said in the [original reply here](https://github.com/mikro-orm/mikro-orm/discussions/2103#discussioncomment-1159626).

If you care about having ts-morph (and therefore the full TS) installed in production, that is a different story.

Using reflect metadata will help with both, but comes with some [limitations](https://mikro-orm.io/docs/metadata-providers/#limitations-and-requirements).

**Reply by derolf** [reply]
https://github.com/mikro-orm/mikro-orm/discussions/2266

Creating self-contained bundle with rollup.
